src/i2c.rs

```diff
@@ -1,579 +1,1819 @@
+        }
+    }
+
+    unsafe impl<'d> Send for AsyncI2cDriver<'d> {}
+
+    impl<'d> Drop for AsyncI2cDriver<'d> {
+        fn drop(&mut self) {
+            loop {
+                if let Ok(_lock_guard) = self.bus_lock.try_lock() {
+                    esp!(unsafe { i2c_del_master_bus(self.handle) }).unwrap();
+                    break;
                 }
             }
         }
+    }
 
-        command_link.master_stop()?;
+    #[cfg(not(esp_idf_i2c_isr_iram_safe))]
+    pub struct AsyncI2cDeviceDriver<'d, T>
+    where
+        T: Borrow<AsyncI2cDriver<'d>>,
+    {
+        driver: T,
+        handle: i2c_master_dev_handle_t,
+        _p: PhantomData<&'d mut ()>,
+    }
 
-        self.cmd_begin(&command_link, timeout)
+    #[cfg(not(esp_idf_i2c_isr_iram_safe))]
+    impl<'d, T> AsyncI2cDeviceDriver<'d, T>
+    where
+        T: Borrow<AsyncI2cDriver<'d>>,
+    {
+        fn new(driver: T, config: &config::DeviceConfig) -> Result<Self, EspError> {
+            let handle = init_device(driver.borrow().bus_handle(), config)?;
+
+            Ok(Self {
+                driver,
+                handle,
+                _p: PhantomData,
+            })
+        }
     }
 
-    fn cmd_begin(
-        &mut self,
-        command_link: &CommandLink,
-        timeout: TickType_t,
-    ) -> Result<(), EspError> {
-        esp!(unsafe { i2c_master_cmd_begin(self.port(), command_link.0, timeout) })
+    #[cfg(not(esp_idf_i2c_isr_iram_safe))]
+    impl<'d, T> AsyncI2cDeviceDriver<'d, T>
+    where
+        T: Borrow<AsyncI2cDriver<'d>>,
+    {
+        pub async fn read(
+            &mut self,
+            buffer: &mut [u8],
+            timeout: TickType_t,
+        ) -> Result<(), EspError> {
+            let handle = self.handle;
+            let driver = self.driver.borrow();
+            let port = driver.port();
+
+            let _lock_guard = driver.acquire_bus().await;
+            enable_master_dev_isr_callback(handle, port)?;
+            esp!(unsafe {
+                i2c_master_receive(
+                    handle,
+                    buffer.as_mut_ptr().cast(),
+                    buffer.len(),
+                    timeout as i32,
+                )
+            })?;
+
+            NOTIFIER[port as usize].wait().await;
+            disable_master_dev_isr_callback(handle)?;
+            Ok(())
+        }
+
+        pub async fn write(&mut self, bytes: &[u8], timeout: TickType_t) -> Result<(), EspError> {
+            let handle = self.handle;
+            let driver = self.driver.borrow();
+            let port = driver.port();
+
+            let _lock_guard = driver.acquire_bus().await;
+            enable_master_dev_isr_callback(handle, port)?;
+            esp!(unsafe {
+                i2c_master_transmit(handle, bytes.as_ptr().cast(), bytes.len(), timeout as i32)
+            })?;
+
+            NOTIFIER[port as usize].wait().await;
+            disable_master_dev_isr_callback(handle)?;
```

Comment:
It is a bit unclear to me from the ESP IDF docs whether the `transmit` function can **still** return with `ESP_ERR_TIMEOUT` when you have the callback registered? Maybe we need to examine the source code?

Because if that's the case, then we need a more complex logic here, potentially involving a `loop`?